### PR TITLE
First-pass ledger hashes in replayer checkpoint files

### LINKED ISSUE #13732: Save pending ledger hashes to checkpoint files
#13715 added a check for SNARKed ledger hashes, that there exists some transaction that produced each such hash.

If replaying from a checkpoint, the transaction that produced a SNARKed ledger hash may have been replayed in an earlier run, so it's not available.

Add the pending ledger hashes to checkpoint files, so those hashes can be considered.

### PR DESCRIPTION
**Pull request by psteckler**
The replayer checks that every new snarked ledger hash it sees is some first-pass ledger hash. If we run the replayer from a checkpoint file, a new snarked ledger hash might be equal to a first-pass ledger hash that was produced before the checkpoint.

Solution: save the pending first-pass ledger hashes to the checkpoint file, and populate the set of hashes from the saved hashes when the replayer starts. Also save the last snarked ledger hash, which is used to know whether to search the first-pass hashes.

The set of ledger hashes are indexed by order of occurrence. That indexing allows flushing older hashes when we find the first-pass hash that matches a snarked ledger hash. The ledger hashes are saved as a list whose order is given by the indexes. When starting the replayer, the set of hashes uses new indexes created when iterating over that list.

The `first_pass_ledger_hashes` and `last_snarked_ledger_hash` fields are optional in the input file, so compatibility with old checkpoint files is maintained. Confirmed that fact by running the replayer on an old checkpoint file.

Closes #13732.

### CODE REVIEW

**Comment by psteckler** [comment]
!ci-build-me

**Comment by psteckler** [comment]
!ci-build-me

**Comment by psteckler** [comment]
!ci-build-me

**Review by dkijania — APPROVED** [review]

File: src/app/replayer/replayer.ml [hunk 1]
```diff
@@ -62,6 +64,42 @@ let create_ledger_as_list ledger =
...
+  let hash_set = Hash_set.create ()
+
+  let add =
+    let count = ref 0 in
+    fun ledger_hash ->
+      Base.Hash_set.add hash_set (ledger_hash, !count) ;
+      incr count
+
+  let find ledger_hash =
+    Base.Hash_set.find hash_set ~f:(fun (hash, _n) ->
```
Comment (georgeee) [inline src/app/replayer/replayer.ml:87]: This is an `O(n)` operation, right?

If yes, I wonder why the hash set is used instead of simple list?

**Review by georgeee — APPROVED** [review]

**Comment by deepthiskumar** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!approved-for-mainnet

File: src/app/replayer/replayer.ml (hunk 1 above)
Comment (psteckler) [reply src/app/replayer/replayer.ml:87]: Yes. Probably when I started on it, I thought we could do lookups on the `hash` alone.

The hash set is a mutable data structure, with the list we'd need to use a mutable reference. Not sure which is better.

**Pull request merged**